utilities/write_batch_with_index/write_batch_with_index_internal.cc

```diff
@@ -151,34 +153,105 @@ int WriteBatchEntryComparator::CompareKey(uint32_t column_family,
+  // this is guaranteed with Valid()
+  assert(iter_entry != nullptr &&
+         iter_entry->column_family == column_family_id_);
+  auto s = write_batch_->GetEntryFromDataOffset(
+      iter_entry->offset, &ret.type, &ret.key, &ret.value, &blob, &xid);
+  assert(s.ok());
+  assert(ret.type == kPutRecord || ret.type == kDeleteRecord ||
+         ret.type == kSingleDeleteRecord || ret.type == kDeleteRangeRecord ||
+         ret.type == kMergeRecord);
+  return ret;
+}
+
+bool WBWIIteratorImpl::MatchesKey(uint32_t cf_id, const Slice& key) {
+  if (Valid()) {
+    return comparator_->CompareKey(cf_id, key, Entry().key) == 0;
+  } else {
+    return false;
+  }
+}
+
+WriteBatchWithIndexInternal::WriteBatchWithIndexInternal(
+    DB* db, ColumnFamilyHandle* column_family)
+    : db_(db), db_options_(nullptr), column_family_(column_family) {
+  if (db_ != nullptr && column_family_ == nullptr) {
+    column_family_ = db_->DefaultColumnFamily();
+  }
+}
+
+WriteBatchWithIndexInternal::WriteBatchWithIndexInternal(
+    const DBOptions* db_options, ColumnFamilyHandle* column_family)
+    : db_(nullptr), db_options_(db_options), column_family_(column_family) {}
+
+WriteBatchWithIndexInternal::WriteBatchWithIndexInternal(
+    ColumnFamilyHandle* column_family)
+    : db_(nullptr), db_options_(nullptr), column_family_(column_family) {}
+
+Status WriteBatchWithIndexInternal::MergeKey(const Slice& key,
+                                             const Slice* value,
+                                             MergeContext& merge_context,
+                                             std::string* result,
+                                             Slice* result_operand) {
+  if (column_family_ != nullptr) {
+    auto cfh = reinterpret_cast<ColumnFamilyHandleImpl*>(column_family_);
+    const auto merge_operator = cfh->cfd()->ioptions()->merge_operator;
+    if (merge_operator == nullptr) {
+      return Status::InvalidArgument(
+          "Merge_operator must be set for column_family");
+    } else if (db_ != nullptr) {
+      const ImmutableDBOptions& immutable_db_options =
+          static_cast_with_check<DBImpl>(db_->GetRootDB())
+              ->immutable_db_options();
+      Statistics* statistics = immutable_db_options.statistics.get();
+      Env* env = immutable_db_options.env;
+      Logger* logger = immutable_db_options.info_log.get();
+
+      return MergeHelper::TimedFullMerge(
+          merge_operator, key, value, merge_context.GetOperands(), result,
+          logger, statistics, env, result_operand);
+    } else if (db_options_ != nullptr) {
+      Statistics* statistics = db_options_->statistics.get();
+      Env* env = db_options_->env;
+      Logger* logger = db_options_->info_log.get();
+      return MergeHelper::TimedFullMerge(
+          merge_operator, key, value, merge_context.GetOperands(), result,
+          logger, statistics, env, result_operand);
+    } else {
+      return MergeHelper::TimedFullMerge(
```

Comment:
I don't think this is a valid state in which to call `TimedFullMerge`. If you call `TimedFullMerge` with a `nullptr` as the `Logger`. Then the MergeOperator implementation may SIGSEGV because AFAIK, Logger should not be null when calling `MergeOperator::FullMergeV2` or `MergeOperator::FullMerge`.

Reply:
Hmm perhaps not... It seems that its perfectly valid for the `Logger*` to be null. So don't worry ;-)